merlin/analysis/filterbarcodes.py

```diff
@@ -57,118 +58,301 @@ def _run_analysis(self, fragmentIndex):
+                         right=True),
+             np.digitize(selectData[:, 2], self.get_area_bins()))) - 1
+        barcodeBins[0, :] = np.clip(
+            barcodeBins[0, :], 0, blankFractionHistogram.shape[0]-1)
+        barcodeBins[1, :] = np.clip(
+            barcodeBins[1, :], 0, blankFractionHistogram.shape[1]-1)
+        barcodeBins[2, :] = np.clip(
+            barcodeBins[2, :], 0, blankFractionHistogram.shape[2]-1)
+        raveledIndexes = np.ravel_multi_index(
+            barcodeBins[:, :], blankFractionHistogram.shape)
 
         thresholdedBlankFraction = blankFractionHistogram < blankThreshold
         return barcodeSet[np.take(thresholdedBlankFraction, raveledIndexes)]
 
-    def _calculate_error_rate(self, barcodeSet):
-        codebook = self.get_codebook()
-        blankFraction = len(
-            codebook.get_blank_indexes()) / codebook.get_barcode_count()
-        return np.sum(barcodeSet['barcode_id'].isin(
-            codebook.get_blank_indexes())) / (len(barcodeSet) * blankFraction)
+    @staticmethod
+    def _extract_counts(barcodes, intensityBins, distanceBins, areaBins):
+        barcodeData = barcodes[
+            ['mean_intensity', 'min_distance', 'area']].values
+        barcodeData[:, 0] = np.log10(barcodeData[:, 0])
+        return np.histogramdd(
+            barcodeData, bins=(intensityBins, distanceBins, areaBins))[0]
 
     def _run_analysis(self):
         decodeTask = self.dataSet.load_analysis_task(
             self.parameters['decode_task'])
-        codebook = self.get_codebook()
+        codebook = decodeTask.get_codebook()
+        barcodeDB = decodeTask.get_barcode_database()
+
+        completeFragments = \
+            self.dataSet.load_numpy_analysis_result_if_available(
+                'complete_fragments', self, [False]*self.fragment_count())
+        pendingFragments = [
+            decodeTask.is_complete(i) and not completeFragments[i]
+            for i in range(self.fragment_count())]
+
+        areaBins = self.dataSet.load_numpy_analysis_result_if_available(
+            'area_bins', self, np.arange(1, 35))
+        distanceBins = self.dataSet.load_numpy_analysis_result_if_available(
+            'distance_bins', self,
+            np.arange(
+                0, decodeTask.parameters['distance_threshold']+0.02, 0.01))
+        intensityBins = self.dataSet.load_numpy_analysis_result_if_available(
+            'intensity_bins', self, None)
+
+        blankCounts = self.dataSet.load_numpy_analysis_result_if_available(
+            'blank_counts', self, None)
+        codingCounts = self.dataSet.load_numpy_analysis_result_if_available(
+            'coding_counts', self, None)
+
+        self.dataSet.save_numpy_analysis_result(
+            areaBins, 'area_bins', self)
+        self.dataSet.save_numpy_analysis_result(
+            distanceBins, 'distance_bins', self)
+
+        updated = False
+        while not all(completeFragments):
```

Comment:
I'm trying to think about what happens here if you start this task after all the decoding is finished. It would seem like completedFragments would return all True, and then it wouldn't enter the loop. Is this intended behavior?